File: electron/tests/user-data_test.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// THE ONE-TIME MOVE of the user-data directory.
//
// Everything a player owns is in that folder — the roster in `localStorage`,
// their settings, the mods they dropped in — so the rules for moving it are
// worth pinning exactly. The decision is a pure function precisely so it can be
// tested without an Electron runtime.

import { describe, expect, it } from "vitest";

import {
  APP_DIR_NAME,
  LEGACY_DIR_NAMES,
  planUserDataMove,
} from "../src/user-data";

const APPDATA = "/home/player/.config";
const at = (name: string) => `${APPDATA}/${name}`;
/** `exists` for exactly this set of directories. */
const only =
  (...dirs: string[]) =>
  (dir: string) =>
    dirs.includes(dir);

describe("planning the move", () => {
  it("moves an install that still carries the npm package's name", () => {
    expect(planUserDataMove(APPDATA, only(at("adas-trail-desktop")))).toEqual({
      from: at("adas-trail-desktop"),
      to: at(APP_DIR_NAME),
    });
  });

  it("moves a macOS install named after the bundle", () => {
    expect(planUserDataMove(APPDATA, only(at("Adas Trail")))).toEqual({
      from: at("Adas Trail"),
      to: at(APP_DIR_NAME),
    });
  });

  it("does nothing for a fresh install", () => {
    expect(planUserDataMove(APPDATA, only())).toBeNull();
  });

  it("does nothing once the move has already happened", () => {
    expect(planUserDataMove(APPDATA, only(at(APP_DIR_NAME)))).toBeNull();
  });

  it("never overwrites data that is already in the new folder", () => {
    // Both present: the new one wins and the old is left where it is. Merging
    // two rosters, or clobbering one with the other, is exactly the wrong
    // thing to guess at — and the old folder can still be recovered by hand.
    const both = only(at(APP_DIR_NAME), at("adas-trail-desktop"));
    expect(planUserDataMove(APPDATA, both)).toBeNull();
  });

  it("prefers the first legacy name when an install somehow has two", () => {
    const both = only(at("adas-trail-desktop"), at("Adas Trail"));
    expect(planUserDataMove(APPDATA, both)?.from).toBe(at(LEGACY_DIR_NAMES[0]));
  });

  it("is named after the executable, with no punctuation to quote", () => {
    // The folder, the binary and the docs all say the same word — and an
    // apostrophe can never reach a path through it.
    expect(APP_DIR_NAME).toBe("adastrail");
    expect(APP_DIR_NAME).toMatch(/^[a-z0-9]+$/);
  });
});

File: tauri/shell/tests/user_data_test.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
//! WHERE THE PLAYER'S THINGS LIVE, and the one time that moves.
//!
//! The move is conservative in both directions, and both directions are a way
//! to lose somebody's window layout, mods and launch history: moving onto a
//! folder that already exists overwrites a fresh install, and merging two would
//! mean guessing which one is real.

use std::path::{Path, PathBuf};

use adastrail_shell::user_data::{
    plan_user_data_move, user_data_dir, UserDataMove, APP_DIR_NAME, LEGACY_DIR_NAMES,
};

/// A fake filesystem: the set of directories that exist.
fn existing(paths: &[PathBuf]) -> impl Fn(&Path) -> bool + '_ {
    move |path: &Path| paths.iter().any(|candidate| candidate == path)
}

#[test]
fn a_fresh_install_moves_nothing() {
    let root = Path::new("/appdata");
    assert_eq!(plan_user_data_move(root, &existing(&[])), None);
}

#[test]
fn an_install_that_already_has_the_folder_is_left_alone() {
    // Either a fresh install or a migration that already happened. Both are
    // "do not touch".
    let root = Path::new("/appdata");
    let ours = user_data_dir(root);
    assert_eq!(plan_user_data_move(root, &existing(&[ours])), None);
}

#[test]
fn the_folder_is_the_declared_name_rather_than_the_bundle_identifier() {
    // Tauri would name it after the reverse-domain identifier, which nobody has
    // ever seen. The executable's own name is the one the docs can print.
    assert_eq!(
        user_data_dir(Path::new("/appdata")),
        Path::new("/appdata").join(APP_DIR_NAME)
    );
    assert!(
        !APP_DIR_NAME.contains('.'),
        "a path segment, not a bundle id"
    );
    assert!(
        !APP_DIR_NAME.contains(' ') && !APP_DIR_NAME.contains('\''),
        "an apostrophe must never reach a path"
    );
}

#[test]
fn the_folder_is_the_executables_name_and_the_old_one_is_adopted() {
    // One desktop wrapper is left, so the folder is the executable's own name —
    // and the name this shell used while there were two is a legacy name, so
    // an install that ran under it is moved rather than orphaned.
    assert_eq!(APP_DIR_NAME, "adastrail");
    assert!(LEGACY_DIR_NAMES.contains(&"adastrail-tauri"));
}

#[test]
fn a_legacy_folder_is_adopted_rather_than_orphaned() {
    // Drives the planner through the same seam the rename uses.
    let root = Path::new("/appdata");
    let Some(legacy) = LEGACY_DIR_NAMES.first() else {
        assert!(
            plan_user_data_move(root, &existing(&[root.join("anything-else")])).is_none(),
            "with no legacy names, nothing is ever adopted"
        );
        return;
    };
    let from = root.join(legacy);
    let planned = plan_user_data_move(root, &existing(std::slice::from_ref(&from)));
    assert_eq!(
        planned,
        Some(UserDataMove {
            from,
            to: user_data_dir(root)
        })
    );
}
